palantir-java-format/src/main/java/com/palantir/javaformat/doc/State.java

```diff
@@ -16,93 +16,204 @@
 
 package com.palantir.javaformat.doc;
 
-import com.google.common.base.MoreObjects;
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableMap;
+import com.google.errorprone.annotations.Immutable;
 import com.palantir.javaformat.Indent;
+import fj.P;
+import fj.P2;
+import fj.data.List;
+import fj.data.Set;
+import fj.data.TreeMap;
+import org.immutables.value.Value;
+import org.immutables.value.Value.Lazy;
+import org.immutables.value.Value.Parameter;
 
 /** State for writing. */
-public final class State {
+@Value.Immutable
+@Value.Style(overshadowImplementation = true)
+@Immutable
+public abstract class State {
     /** Last indent that was actually taken. */
-    final int lastIndent;
+    public abstract int lastIndent();
     /** Next indent, if the level is about to be broken. */
-    final int indent;
+    public abstract int indent();
 
-    final int column;
-    final boolean mustBreak;
+    public abstract int column();
+
+    public abstract boolean mustBreak();
     /** Counts how many lines a particular formatting took. */
-    final int numLines;
+    public abstract int numLines();
     /**
      * Counts how many times reached a branch, where multiple formattings would be considered. Expected runtime is
      * exponential in this number.
      *
      * @see State#withNewBranch()
      */
-    final int branchingCoefficient;
+    public abstract int branchingCoefficient();
+
+    protected abstract Set<BreakTag> breakTagsTaken();
+
+    protected abstract TreeMap<Break, BreakState> breakStates();
+
+    protected abstract TreeMap<Level, LevelState> levelStates();
+
+    protected abstract List<P2<Tok, TokState>> tokStates();
```

Comment:
wanna justify why this isn't a TreeMap too?

Reply:
I thought this is a perf optimization since they are all first written and then only the final state is read as a map, but actually these are only comments, so there aren't that many.
Therefore, I've made this a TreeMap too.